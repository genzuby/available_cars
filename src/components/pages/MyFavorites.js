import React from 'react';
import { Container } from 'react-bootstrap';
import FavoriteCars from '../Content/FavoriteCars';
import { mainStyle } from '../../style/componentStyle';

function MyFavorites() {
  return (
    <Container
      fluid
      className="d-flex"
      style={mainStyle}
      data-testid="myFavorites"
    >
      <FavoriteCars />
    </Container>
  );
}

export default MyFavorites;
