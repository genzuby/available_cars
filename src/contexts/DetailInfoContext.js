import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getDetailInfo } from '../redux/actions';

export const DetailInfoContext = React.createContext();

export function DetailInfoProvider({ children }) {
  const [detailInfo, setDetailInfo] = useState({});
  const carDetailInfo = useSelector(state => state.carDetailInfo);
  const dispatch = useDispatch();

  useEffect(() => {
    setDetailInfo(carDetailInfo);
  }, [carDetailInfo]);

  const fetchDetailInfo = stockNumber => {
    dispatch(getDetailInfo(stockNumber));
  };

  return (
    <DetailInfoContext.Provider value={{ detailInfo, fetchDetailInfo }}>
      {children}
    </DetailInfoContext.Provider>
  );
}
