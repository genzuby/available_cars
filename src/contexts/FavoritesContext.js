import React from 'react';
import { useSelector } from 'react-redux';
import useLocalStorage from '../hooks/useLocalStorage';

export const FavoritesContext = React.createContext();

const LOCAL_STOAGE_KEY = 'AUTO1_MY_FAVORITE_CAR';

export function FavoritesProvider({ children }) {
  const [favorites, setFavorites] = useLocalStorage(LOCAL_STOAGE_KEY, []);
  const carDetailInfo = useSelector(state => state.carDetailInfo);

  const addFavorite = () => {
    setFavorites(prevFavorites => {
      let checkDuplication;
      if (prevFavorites !== 0) {
        checkDuplication = prevFavorites.filter(
          item => item.stockNumber === carDetailInfo.stockNumber
        );
      }

      return checkDuplication.length !== 0
        ? prevFavorites
        : [carDetailInfo, ...prevFavorites];
    });
  };

  const deleteFavorite = (stockNumber: string) => {
    setFavorites(prevFavorites => {
      const filteredFavorites = prevFavorites.filter(
        item => item.stockNumber !== +stockNumber
      );

      return filteredFavorites;
    });
  };

  const value = { favorites, addFavorite, deleteFavorite };

  return (
    <FavoritesContext.Provider value={value}>
      {children}
    </FavoritesContext.Provider>
  );
}
