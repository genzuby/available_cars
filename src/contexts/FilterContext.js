import React, { useState } from 'react';

export const FilterContext = React.createContext();

export function FilterProvider({ children }) {
  const [filter, setFilter] = useState({});

  return (
    <FilterContext.Provider value={{ filter, setFilter }}>
      {children}
    </FilterContext.Provider>
  );
}
