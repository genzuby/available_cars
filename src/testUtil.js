import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

export const providerValue = {
  color: 'red',
  fuelType: 'Diesel',
  manufacturerName: 'Dodge',
  mileage: { number: 138210, unit: 'km' },
  modelName: 'Caliber',
  pictureUrl: 'car.svg',
  stockNumber: 10907,
};

const initialState = {
  carDetailInfo: providerValue,
  manufacturersFilter: [{ name: 'Audi' }],
  colorFilter: ['red'],
  carList: { cars: [providerValue], totalPageCount: 10, totalCarsCount: 100 },
};

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

export const store = mockStore(initialState);

export const localStorageMock = {
  getItem: jest.fn().mockImplementation(key => localStorageItems[key]),
  setItem: jest.fn().mockImplementation((key, value) => {
    localStorageItems[key] = value;
  }),
  clear: jest.fn().mockImplementation(() => {
    localStorageItems = {};
  }),
  removeItem: jest.fn().mockImplementation(key => {
    localStorageItems[key] = undefined;
  }),
};

export let localStorageItems = {};
