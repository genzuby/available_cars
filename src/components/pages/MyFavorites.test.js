import { render, screen } from '@testing-library/react';
import MyFavorites from './MyFavorites';
import { FavoritesProvider } from '../../contexts/FavoritesContext';
import { Provider } from 'react-redux';
import { store } from '../../testUtil';

beforeEach(() => {
  render(
    <Provider store={store}>
      <FavoritesProvider value={{ favorites: {} }}>
        <MyFavorites />
      </FavoritesProvider>
    </Provider>
  );
});

it('snapshot', () => {
  expect(screen.getByTestId('myFavorites')).toMatchSnapshot();
});

it('renders with context value', () => {
  expect(screen.getByTestId('myFavorites')).toBeInTheDocument();
});
