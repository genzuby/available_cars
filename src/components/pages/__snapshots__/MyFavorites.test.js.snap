// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="d-flex container-fluid"
  data-testid="myFavorites"
  style="margin: 80px 0px; padding: 24px 48px 0px 24px; display: grid; min-height: 76vh;"
>
  <div
    class="sc-hKgILt kUrLvW"
    data-testid="favoriteCars"
  >
    <h3
      class="--list-title"
    >
      My Fovorite cars
    </h3>
    <p
      class="--list-desc"
    >
      0
       results
    </p>
    <div
      data-testid="carList"
    />
  </div>
</div>
`;
